Ed (edwards curve), EdDSA (Ed25519/Ed448) and XDH (X25519/X448)

ariawisp opened the pull request:
This is a rebased version of https://github.com/whyoleg/cryptography-kotlin/pull/61 that also addresses the outstanding comments on the original PR, and adds Apple providers.

whyoleg commented on cryptography-providers/webcrypto/src/commonMain/kotlin/internal/Algorithms.kt:
```diff
@@ -31,6 +31,7 @@ internal expect val Algorithm.ecKeyAlgorithmNamedCurve: String
 internal expect fun EcdsaSignatureAlgorithm(hash: String): Algorithm
 
 internal expect fun EcdhKeyDeriveAlgorithm(publicKey: CryptoKey): Algorithm
```
it looks like we can just remove `EcdhKeyDeriveAlgorithm` and use `KeyDeriveAlgorithm` everywhere?

whyoleg reviewed (changes requested):
Nice! Thanks for the PR! I've performed first round of review and left quite a lot of comments. Hope that most of them could be easily solved!

P.S. I'm quite curious, was any AI tool used to implement any part of the PR? There is no problem with it, just curiocity :)

ariawisp commented:
Thanks for the review @whyoleg ! All the comments should be addressed now.

To answer your question regarding AI: yes, OpenAI's Codex CLI with GPT-5 (high reasoning) was used for the entirety of this PR.

Highlights of changes since your review:
- Docs:
  - jdk.md/openssl3.md: removed extra “EdDSA/XDH” sections to avoid duplication.
  - cryptokit.md: clarified only Ed25519/X25519 supported; DER/PEM and RAW encodings available via provider APIs.
  - webcrypto.md: clarified EdDSA/XDH were added later to WebCrypto; added RFC references (RFC 8032, RFC 7748).

- WebCrypto:
  - Removed engine detection and all gating; rely on runtime availability errors from the engine.
  - Implemented JWK and RAW support for EdDSA/XDH (public/private) with PEM/DER wrapping via processors.
  - Unified key derivation to use KeyDeriveAlgorithm (removed the separate ECDH variant).

- OpenSSL3:
  - Added RAW key format support for EdDSA/XDH via SPKI/PKCS#8 wrap/unwrap on encode/decode.
  - Extracted shared-secret derivation into operations and reused it.
  - Switched EdDSA to one‑shot mode (nullable hashAlgorithm) in digest-based sign/verify.

- CryptoKit:
  - Moved Ed25519/X25519 OIDs into asn1-modules and used them.
  - Extracted accumulating sign/verify into provider-base (AccumulatingSignFunction/VerifyFunction) and adopted in EdDSA.
  - Compatibility tests skip only unsupported curves (Ed448/X448).

- JDK:
  - Added RAW support to EdDSA (via SPKI/PKCS#8 wrap/unwrap).
  - Tests now reflect correct split: XDH since JDK 11, EdDSA since JDK 15.

- Tests:
  - Fixed provider checks to use context.provider where appropriate.
  - Simplified SupportedAlgorithmsTest gating (Apple vs others).
  - Removed WebCrypto browser skip rules; let engines handle availability at runtime.

- ASN.1:
  - AlgorithmIdentifier deserializer now tolerates absent/NULL parameters (RFC 8410) to avoid decode errors on SPKI/PKCS#8.
  - Added Edwards/Montgomery OIDs to asn1-modules.

whyoleg commented on cryptography-providers/cryptokit/src/commonMain/kotlin/algorithms/CryptoKitEdDSA.kt:
```diff
@@ -0,0 +1,140 @@
+        override val privateKey: EdDSA.PrivateKey,
+    ) : EdDSA.KeyPair
+
+    private class EdPublic(
```
should be `EdPublicKey`

whyoleg commented on cryptography-providers/cryptokit/src/commonMain/kotlin/algorithms/CryptoKitEdDSA.kt:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    private class EdPrivate(
```
should be `EdPrivateKey`

whyoleg reviewed (changes requested):
The PR is really big so reviewing it becoming a bit hard, but here is the second round of review.

If the whole PR is done via AI - I will probably need some time more to review it locally and perform some additional manual changes, to split commits/changes, so that the history will be nicer. Or you can do it if you like :)
Depending on your answer, I will provide more instructions

ariawisp commented on swiftinterop/src/main/kotlin/tasks/GenerateSwiftPackageDefinitionTask.kt:
```diff
@@ -69,7 +70,11 @@ abstract class GenerateSwiftPackageDefinitionTask : DefaultTask() {
                     .target(
-                        name: "$swiftinteropModuleName"
+                        name: "$swiftinteropModuleName",
+                        swiftSettings: [
```
These changes were made to address linker/compiler issues on macOS 26 beta  / Xcode 26.

ariawisp commented on swiftinterop/src/main/kotlin/tasks/GenerateSwiftPackageDefinitionTask.kt:
```diff
@@ -44,11 +44,12 @@ abstract class GenerateSwiftPackageDefinitionTask : DefaultTask() {
         outputDirectory.get().asFile.recreateDirectories()
 
         val swiftinteropModuleName = swiftinteropModuleName.get()
+        fun ver(v: String) = ".v$v"
```
These changes were made to address linker/compiler issues on macOS 26 beta  / Xcode 26.

ariawisp commented on swiftinterop/src/main/kotlin/tasks/XcodebuildBuildTask.kt:
```diff
@@ -45,13 +45,19 @@ abstract class XcodebuildBuildTask : DefaultTask() {
         exec.exec {
             it.workingDir(temporaryDir)
-            it.commandLine(
+            val developerDir = System.getenv("DEVELOPER_DIR") ?: ""
```
These changes were made to address linker/compiler issues on macOS 26 beta  / Xcode 26.

ariawisp commented on swiftinterop/src/main/kotlin/tasks/GenerateSwiftCinteropDefinitionTask.kt:
```diff
@@ -75,7 +75,8 @@ abstract class GenerateSwiftCinteropDefinitionTask : DefaultTask() {
         version: Provider<String>,
     ): String {
-        val linker = "-L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/$libsDir/"
+        val developerDir = System.getenv("DEVELOPER_DIR") ?: "/Applications/Xcode.app/Contents/Developer"
```
These changes were made to address linker/compiler issues on macOS 26 beta  / Xcode 26.

ariawisp commented:
Thanks for the second review @whyoleg - let me see what I can do for breaking this up into more smaller PRs

ariawisp commented:
# Split PRs Overview: EdDSA/XDH Series

Below is a concise list of the new PRs that split the original combined work, with links, scope, and dependencies. All target `main` and keep test/CI setup unchanged. Recommended merge order is Foundation → Providers → Tests → Docs.

## Merge Order (high level)
- Foundation: #108 → #109 → #110
- Providers: #111 (JDK), #112 (OpenSSL3), #113 (WebCrypto), #114 (CryptoKit)
- Tests: #115
- Docs: #116, #117, #118 (per provider)

## Foundation
- #108 — ASN.1: Ed25519/Ed448 and X25519/X448 OIDs; accept ABSENT/NULL AlgorithmIdentifier params (RFC 8410)
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/108
  - Branch: `feat/asn1-oids-rfc8410`
  - What: expose OIDs as `ObjectIdentifier` extensions; tolerate ABSENT/NULL params; add minimal decode tests; ABI updated.

- #109 — Core: Add EdDSA and XDH algorithm interfaces
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/109
  - Branch: `feat/core-ed-xdh`
  - What: add `EdDSA`/`XDH` APIs and core API descriptors (no provider/test changes).

- #110 — Providers/Base: Add accumulating sign/verify helpers
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/110
  - Branch: `feat/provider-base-accum`
  - What: add `AccumulatingSignFunction`/`AccumulatingVerifyFunction` (mirrors cipher helper); add shared `unwrapSubjectPublicKeyInfo(Set<ObjectIdentifier>)`; ABI updated.

## Providers
- #111 — JDK: Implement EdDSA and XDH; RAW key wrap/unwrap helper; wire provider
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/111
  - Branch: `feat/jdk-ed-xdh`
  - What: JDK EdDSA/XDH; uses base SPKI unwrap helper; keeps minimal PKCS#8 unwrap helper; uses `ObjectIdentifier.*` OIDs; provider wiring.
  - Depends on: Core #109 (uses APIs); references Base #110; uses ASN.1 OIDs from #108 for DER/PEM.

- #112 — OpenSSL3: Implement EdDSA and XDH; refactor key agreement; wire provider
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/112
  - Branch: `feat/openssl3-ed-xdh`
  - What: OpenSSL3 EdDSA/XDH; refactors key agreement; one‑shot uses accumulating helpers; uses `ObjectIdentifier.*` OIDs; provider wiring.
  - Depends on: Core #109; references ASN.1 #108 (DER/PEM OIDs).

- #113 — WebCrypto: Implement EdDSA and XDH; internal algorithms; wire provider
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/113
  - Branch: `feat/webcrypto-ed-xdh`
  - What: WebCrypto EdDSA/XDH; internal algorithms mapping (common/js/wasm); uses `ObjectIdentifier.*` OIDs; provider wiring.
  - Depends on: Core #109; references ASN.1 #108 where DER/PEM OIDs are used.

- #114 — CryptoKit: Implement EdDSA and XDH; add Swift sources; wire provider
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/114
  - Branch: `feat/cryptokit-ed-xdh`
  - What: CryptoKit Ed25519/X25519; decode unwrap‑to‑RAW + `decodeRawWithRawRepresentation`; encode hoists raw; inner classes renamed to `*Key`; uses `ObjectIdentifier.*` OIDs; provider wiring.
  - Depends on: Core #109; references ASN.1 #108 (DER/PEM OIDs).

## Tests
- #115 — Tests: Add EdDSA/XDH default and compatibility suites; unify capability gating
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/115
  - Branch: `tests/ed-xdh`
  - What: add default + compatibility tests for EdDSA/XDH; consolidate gating in `support.kt`; update generator. No runner/CI changes.
  - Depends on: Core #109 to compile on CI.

## Docs (per provider)
- #116 — Docs(CryptoKit): Add EdDSA/XDH capability note (Ed25519/X25519 only)
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/116
  - Branch: `docs/cryptokit-ed-xdh`

- #117 — Docs(WebCrypto): Add EdDSA/XDH availability note and RFC links
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/117
  - Branch: `docs/webcrypto-ed-xdh`

- #118 — Docs(JDK): Add EdDSA/XDH availability note
  - Link: https://github.com/whyoleg/cryptography-kotlin/pull/118
  - Branch: `docs/jdk-ed-xdh`

## Notes
- All PRs target `main` with focused scope and no test/CI setup changes.
- Providers and tests are blocked by Core (#109) to compile; ASN.1 (#108) provides OIDs used in DER/PEM code paths.
- Provider Base (#110) centralizes SPKI unwrap helper and accumulating helpers; providers already updated to use these.

ariawisp commented:
I don't mind at all, please do!
